File: src/com/medium/GenerateParentheses_22.java
/**
 * 
 */
package com.medium;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class GenerateParentheses_22 {
    // 2019年8月9日
    // Use DFS
    // Time Complexity: O(2^n)
    // Space Complexity:O(k + n)
    // Runtime: 1 ms, faster than 93.54%
    // Memory Usage: 36.1 MB, less than 100.00%
    public List<String> generateParenthesis(int n) {
        List<String> mResult = new ArrayList<>();

        if (n <= 0) {
            return mResult;
        }
        helper(n, 0, 0, "", mResult);

        return mResult;
    }

    private void helper(int n, int mCountLeft, int mCountRight, String mTmpStr, List<String> mResult) {
        if (mTmpStr.length() == 2 * n) {
            mResult.add(mTmpStr);
        }

        if (mCountLeft < n) {
            helper(n, mCountLeft + 1, mCountRight, mTmpStr + "(", mResult);
        }
        if (mCountRight < mCountLeft) {
            helper(n, mCountLeft, mCountRight + 1, mTmpStr + ")", mResult);
        }
    }
}

File: Golang_src/medium/generateParentheses_22.go
package medium

// Use backtracking:
// Time Complexity: O((4^n)/(n^1/2))
// Space Complexity:O(n)
// Runtime: 4 ms, faster than 14.53%
// Memory Usage: 3.1 MB, less than 26.74%
func generateParenthesis(n int) []string {
	result := []string{}

	var dfs func(int, int, string)
	dfs = func(left, right int, str string) {
		if left < 0 || right < 0 {
			return
		}

		if left == 0 && right == 0 {
			result = append(result, str)
			return
		}
		if left == right {
			dfs(left-1, right, str+"(")
		} else if left < right {
			dfs(left, right-1, str+")")
			dfs(left-1, right, str+"(")
		}
	}

	dfs(n, n, "")

	return result
}
